test_shim.py

```diff
@@ -0,0 +1,225 @@
+    for path in glob.iglob(f"{dir}/**/*.py", recursive=True):
+        path = path[len(dir) : -3].replace(os.sep, ".")
+        if importlib.util.find_spec(path, package) is not None:
+            submodules.add(path)
+
+    return sorted([submod[:-9] for submod in submodules if submod.endswith("__init__")])
+
+
+def test_package(base_name: str, shim_name: str) -> Response:
+    """
+    Imports and recursively tests all submodules of the base and shim package.
+
+    Offloads testing to test_all_replicas
+    """
+    response: Response = {
+        "extra_files": [],
+        "missing_files": [],
+        "attributes": {},
+    }
+
+    base = importlib.import_module(base_name)
+    shim = importlib.import_module(shim_name)
+    replicas = test_all_replicas(base, shim)
+    response["extra_files"] = replicas[0]
+    response["missing_files"] = replicas[1]
+
+    response["attributes"].update(test_module(base, shim, to_skip=response["missing_files"]))
+    for module in find_all_submodules(shim_name):
+        if not module:
+            continue
+        base = importlib.import_module(base_name + module)
+        shim = importlib.import_module(shim_name + module)
+
+        response["attributes"].update(test_module(base, shim, to_skip=response["missing_files"]))
+    return response
+
+
+def main() -> int:
+    parser = argparse.ArgumentParser(
+        description="Test a shim package is identical to the main package."
+    )
+    parser.add_argument("base", help="Primary module.")
+    parser.add_argument("shim", help="Shim module.", nargs="?", default="discord")
+
+    args = parser.parse_args()
+
+    results = test_package(args.base, args.shim)
+    if results.get("extra_files"):
+        logger.warning(f"Extra modules: {results['extra_files']}")
+    if results.get("missing_files"):
+        logger.error(f"Missing modules: {', '.join(results['missing_files'])}")
+    if results.get("attributes"):
+        for module, attributes in results["attributes"].items():
+            if attributes["missing_attributes"]:
+                logger.error(
+                    f"{module} has missing attributes: {', '.join(attributes['missing_attributes'])}"
+                )
+            if attributes["mismatching_attributes"]:
+                logger.error(
+                    f"{module} has mismatching attributes: {', '.join(attributes['mismatching_attributes'])}"
+                )
+
+    if not any(results.values()):
+        logger.info("All tests passed.")
+        return 0
+    return 1
+
+
+if __name__ == "__main__":
+    try:
+        import coloredlogs
+    except ModuleNotFoundError:
+        logging.basicConfig(level=logging.DEBUG, format="{levelname:<8} -  {message}", style="{")
+    else:
+        coloredlogs.install(level="DEBUG", fmt="{levelname:<8} -  {message}", style="{")
+
+    import time
+
+    start = time.time_ns()
+    res = main()
+    end = time.time_ns()
+    print(f"Took {(end - start) / 1e9} seconds.")
```

Comment:
Shouldn't this use logging?

Reply:
IMO, no. I want to be able to redirect stderr and keep this in the console.